File: src/frontend/contabil/lib/app/data/model/contabil_encerramento_exe_det_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:contabil/app/data/model/model_imports.dart';

class ContabilEncerramentoExeDetModel {
	int? id;
	int? idContabilEncerramentoExe;
	int? idContabilConta;
	double? saldoAnterior;
	double? valorDebito;
	double? valorCredito;
	double? saldo;
	ContabilContaModel? contabilContaModel;

	ContabilEncerramentoExeDetModel({
		this.id,
		this.idContabilEncerramentoExe,
		this.idContabilConta,
		this.saldoAnterior,
		this.valorDebito,
		this.valorCredito,
		this.saldo,
		this.contabilContaModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'saldo_anterior',
		'valor_debito',
		'valor_credito',
		'saldo',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Saldo Anterior',
		'Valor Debito',
		'Valor Credito',
		'Saldo',
	];

	ContabilEncerramentoExeDetModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idContabilEncerramentoExe = jsonData['idContabilEncerramentoExe'];
		idContabilConta = jsonData['idContabilConta'];
		saldoAnterior = jsonData['saldoAnterior']?.toDouble();
		valorDebito = jsonData['valorDebito']?.toDouble();
		valorCredito = jsonData['valorCredito']?.toDouble();
		saldo = jsonData['saldo']?.toDouble();
		contabilContaModel = jsonData['contabilContaModel'] == null ? ContabilContaModel() : ContabilContaModel.fromJson(jsonData['contabilContaModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idContabilEncerramentoExe'] = idContabilEncerramentoExe != 0 ? idContabilEncerramentoExe : null;
		jsonData['idContabilConta'] = idContabilConta != 0 ? idContabilConta : null;
		jsonData['saldoAnterior'] = saldoAnterior;
		jsonData['valorDebito'] = valorDebito;
		jsonData['valorCredito'] = valorCredito;
		jsonData['saldo'] = saldo;
		jsonData['contabilContaModel'] = contabilContaModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idContabilEncerramentoExe = plutoRow.cells['idContabilEncerramentoExe']?.value;
		idContabilConta = plutoRow.cells['idContabilConta']?.value;
		saldoAnterior = plutoRow.cells['saldoAnterior']?.value?.toDouble();
		valorDebito = plutoRow.cells['valorDebito']?.value?.toDouble();
		valorCredito = plutoRow.cells['valorCredito']?.value?.toDouble();
		saldo = plutoRow.cells['saldo']?.value?.toDouble();
		contabilContaModel = ContabilContaModel();
		contabilContaModel?.descricao = plutoRow.cells['contabilContaModel']?.value;
	}	

	ContabilEncerramentoExeDetModel clone() {
		return ContabilEncerramentoExeDetModel(
			id: id,
			idContabilEncerramentoExe: idContabilEncerramentoExe,
			idContabilConta: idContabilConta,
			saldoAnterior: saldoAnterior,
			valorDebito: valorDebito,
			valorCredito: valorCredito,
			saldo: saldo,
		);			
	}

	
}

File: backend/contabil/src/model/contabil_encerramento_exe_det_model.py
from src import db
from src.model.contabil_conta_model import ContabilContaModel


class ContabilEncerramentoExeDetModel(db.Model):
    __tablename__ = 'contabil_encerramento_exe_det'

    id = db.Column(db.Integer, primary_key=True)
    saldo_anterior = db.Column(db.Float)
    valor_debito = db.Column(db.Float)
    valor_credito = db.Column(db.Float)
    saldo = db.Column(db.Float)
    id_contabil_encerramento_exe = db.Column(db.Integer, db.ForeignKey('contabil_encerramento_exe_cab.id'))
    id_contabil_conta = db.Column(db.Integer, db.ForeignKey('contabil_conta.id'))

    contabil_conta_model = db.relationship('ContabilContaModel', foreign_keys=[id_contabil_conta])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_contabil_encerramento_exe = data.get('idContabilEncerramentoExe')
        self.id_contabil_conta = data.get('idContabilConta')
        self.saldo_anterior = data.get('saldoAnterior')
        self.valor_debito = data.get('valorDebito')
        self.valor_credito = data.get('valorCredito')
        self.saldo = data.get('saldo')

    def serialize(self):
        return {
            'id': self.id,
            'idContabilEncerramentoExe': self.id_contabil_encerramento_exe,
            'idContabilConta': self.id_contabil_conta,
            'saldoAnterior': self.saldo_anterior,
            'valorDebito': self.valor_debito,
            'valorCredito': self.valor_credito,
            'saldo': self.saldo,
            'contabilContaModel': self.contabil_conta_model.serialize() if self.contabil_conta_model else None,
        }